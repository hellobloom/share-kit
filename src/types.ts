import {AttestationTypeID, HashingLogic} from '@bloomprotocol/attestations-lib'
import {IProof} from 'merkletreejs'

// Request Types

enum Action {
  attestation = 'request_attestation_data',
}

type RequestData = {
  action: Action
  token: string
  url: string
  org_logo_url: string
  org_name: string
  org_usage_policy_url: string
  org_privacy_policy_url: string
  types: (keyof typeof AttestationTypeID)[]
}

enum ErrorCorrectionLevel {
  'L' = 1,
  'M' = 0,
  'Q' = 3,
  'H' = 2,
}

type QROptions = {
  ecLevel: keyof typeof ErrorCorrectionLevel
  size: number
  bgColor: string
  fgColor: string
  hideLogo: boolean
  padding: number
  logoImage?: string
  logoWidth?: number
  logoHeight?: number
  logoOpacity?: number
}

// Response Types

/**
 * Based on IProof from `merkletreejs`, but the data property is a string
 * which should contain the hex string representation of a Buffer for
 * compatibility when serializing / deserializing.
 */
interface IProofShare {
  position: 'left' | 'right'
  data: string
}

/**
 * Represents the data shared by a user, which has been attested on the Bloom Protocol.
 * Receivers of this data can / should verity this data hasn't been tampered with.
 */
interface IVerifiedData {
  /**
   * Blockchain transaction hash which emits the layer2Hash property
   */
  tx: string

  /**
   * Attestation hash that lives on chain and is formed by hashing the merkle
   * tree root hash with a nonce.
   */
  layer2Hash: string

  /**
   * Merkle tree root hash
   */
  rootHash: string

  /**
   * Nonce used to hash the `rootHash` to create the `layer2Hash`
   */
  rootHashNonce: string

  /**
   * Merkle tree leaf proof
   */
  proof: IProofShare[]

  /**
   * The Ethereum network name on which the tx can be found
   */
  stage: 'mainnet' | 'rinkeby' | 'local'

  /**
   * Data node containing the raw verified data that was requested
   */
  target: HashingLogic.IDataNode

  /**
   * Ethereum address of the attester that performed the attestation
   */
  attester: string
}

type ResponseData = {
  /**
   * The Ethereum address of the user sharing their data
   */
  subject: string

  /**
   * Data shared to the receiving endpoint requested by the share-kit QR code.
   * This data can be verified by the receiver via functions in utils.ts.
   */
  data: IVerifiedData[]

  /**
   * Hex string representation of the `data` being keccak256 hashed
   */
  packedData: string

  /**
   * Signature of `packedData` by the user with their mnemonic.
   */
  signature: string

  /**
   * Token that should match the one provided to the share-kit QR code.
   */
  token: string
}

export {Action, RequestData, ErrorCorrectionLevel, QROptions, IProof, IProofShare, IVerifiedData, ResponseData}
