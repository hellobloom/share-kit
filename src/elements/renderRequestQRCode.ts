import {generateId} from './utils'
import {Options, RequestData, ErrorCorrectionLevel} from '../types'
import {BloomLogo} from '../BloomLogo'

const QRCodeImpl = require('qr.js/lib/QRCode')

const defaultOptions: Options = {
  hideLogo: false,
  ecLevel: 'L',
  size: 128,
  bgColor: '#fff',
  fgColor: '#6067f1',
  padding: 0,
}

type CellInfo = {
  top: number
  left: number
  size: number
  color: string
}

const makeDot = (ctx: CanvasRenderingContext2D, info: CellInfo) => {
  const centerX = info.left + info.size / 2
  const centerY = info.top + info.size / 2
  const radius = (info.size / 2) * 0.85

  ctx.save()
  ctx.beginPath()

  ctx.fillStyle = info.color

  ctx.arc(centerX, centerY, radius, 0, 2 * Math.PI, false)
  ctx.fill()

  ctx.closePath()
  ctx.restore()
}

const makeRect = (ctx: CanvasRenderingContext2D, color: string, rect: {x: number; y: number; w: number; h: number}) => {
  ctx.save()
  ctx.beginPath()

  ctx.fillStyle = color
  ctx.strokeStyle = color
  ctx.lineWidth = 1
  ctx.lineJoin = 'round'
  ctx.lineCap = 'round'

  ctx.rect(rect.x, rect.y, rect.w, rect.h)
  ctx.fill()
  ctx.stroke()

  ctx.closePath()
  ctx.restore()
}

enum CornerType {
  none,
  topLeft,
  topRight,
  bottomRight,
  bottomLeft,
}

const makeCorner = (ctx: CanvasRenderingContext2D, info: CellInfo, type: CornerType, part: EyePart) => {
  const outerRadius = part === EyePart.outer ? info.size : info.size / 4
  const innerRadius = outerRadius / 2

  const {top, left} = info
  const bottom = top + info.size
  const right = left + info.size
  const halfSize = info.size / 2

  ctx.save()
  ctx.beginPath()

  ctx.fillStyle = info.color
  ctx.strokeStyle = info.color
  ctx.lineWidth = 1
  ctx.lineJoin = 'round'
  ctx.lineCap = 'round'

  switch (type) {
    case CornerType.topLeft:
      ctx.moveTo(left, bottom)
      ctx.arcTo(left, top, right, top, outerRadius)
      ctx.lineTo(right, top)
      ctx.lineTo(right, bottom)
      ctx.lineTo(right + halfSize, bottom)
      ctx.arcTo(right, bottom, right, bottom + 5, innerRadius)
      ctx.lineTo(right, bottom + 5)
      ctx.lineTo(right, bottom)
      ctx.lineTo(left, bottom)
      break
    case CornerType.topRight:
      ctx.moveTo(left, bottom)
      ctx.lineTo(left, top)
      ctx.arcTo(right, top, right, bottom, outerRadius)
      ctx.lineTo(right, bottom)
      ctx.lineTo(left, bottom)
      ctx.lineTo(left, bottom + halfSize)
      ctx.arcTo(left, bottom, left - halfSize, bottom, innerRadius)
      ctx.lineTo(left - halfSize, bottom)
      ctx.lineTo(left, bottom)
      break
    case CornerType.bottomRight:
      ctx.moveTo(left, bottom)
      ctx.lineTo(left, top)
      ctx.lineTo(left - halfSize, top)
      ctx.arcTo(left, top, left, top - halfSize, innerRadius)
      ctx.lineTo(left, top - halfSize)
      ctx.lineTo(left, top)
      ctx.lineTo(right, top)
      ctx.arcTo(right, bottom, left, bottom, outerRadius)
      ctx.lineTo(left, bottom)
      break
    case CornerType.bottomLeft:
      ctx.moveTo(left, top)
      ctx.lineTo(right, top)
      ctx.lineTo(right, top - halfSize)
      ctx.arcTo(right, top, right + halfSize, top, innerRadius)
      ctx.lineTo(right + halfSize, top)
      ctx.lineTo(right, top)
      ctx.lineTo(right, bottom)
      ctx.arcTo(left, bottom, left, top, outerRadius)
      ctx.lineTo(left, top)
      break
    default:
      break
  }

  ctx.fill()
  ctx.stroke()

  ctx.closePath()
  ctx.restore()
}

class ConnectionType {
  static readonly none = 0
  static readonly left = 1 << 0
  static readonly right = 1 << 1
  static readonly top = 1 << 2
  static readonly bottom = 1 << 4
}

enum EyePart {
  inner,
  outer,
}

const makeEyeBit = (ctx: CanvasRenderingContext2D, info: CellInfo, connectionType: number, part: EyePart) => {
  const centerX = info.left + info.size / 2
  const centerY = info.top + info.size / 2
  const halfSize = info.size / 2

  const isLeft = (connectionType & ConnectionType.left) === ConnectionType.left
  const isRight = (connectionType & ConnectionType.right) === ConnectionType.right
  const isTop = (connectionType & ConnectionType.top) === ConnectionType.top
  const isBottom = (connectionType & ConnectionType.bottom) === ConnectionType.bottom

  if (isLeft !== isRight && isTop !== isBottom) {
    let cornerType = CornerType.none

    if (isBottom && isRight) {
      cornerType = CornerType.topRight
    } else if (isBottom && isLeft) {
      cornerType = CornerType.topLeft
    } else if (isTop && isRight) {
      cornerType = CornerType.bottomRight
    } else if (isTop && isLeft) {
      cornerType = CornerType.bottomLeft
    }

    makeCorner(ctx, info, cornerType, part)
  } else {
    if (isLeft) {
      makeRect(ctx, info.color, {x: centerX, y: info.top, w: halfSize, h: info.size})
    }

    if (isRight) {
      makeRect(ctx, info.color, {x: info.left, y: info.top, w: halfSize, h: info.size})
    }

    if (isTop) {
      makeRect(ctx, info.color, {x: info.left, y: info.top, w: info.size, h: halfSize})
    }

    if (isBottom) {
      makeRect(ctx, info.color, {x: info.left, y: centerY, w: info.size, h: halfSize})
    }
  }
}

const drawCanvas = (canvas: HTMLCanvasElement, data: RequestData, options: Partial<Options>) => {
  const defaultedOptions = {...defaultOptions, ...options}

  const {ecLevel, size, bgColor, fgColor, padding} = defaultedOptions

  const qr = new QRCodeImpl(-1, ErrorCorrectionLevel[ecLevel])
  qr.addData(JSON.stringify(data))
  qr.make()

  const ctx = canvas.getContext('2d')!
  ctx.clearRect(0, 0, canvas.width, canvas.height)

  const scale = window.devicePixelRatio || 1
  const cells: [boolean[]] = qr.modules
  const cellSize = size / cells.length
  // Add 4 to account for:
  // - The 1 extra on top and left of each cell
  // - The 1 extra from the border on top/left
  canvas.height = canvas.width = size * scale + 4
  canvas.style.height = canvas.style.width = `${size}px`
  canvas.style.padding = (100 * padding) / size + '%'
  ctx.scale(scale, scale)

  makeRect(ctx, bgColor, {x: 0, y: 0, w: size, h: size})

  cells.forEach((row: boolean[], rowIndex: number) => {
    row.forEach((cell: boolean, cellIndex: number) => {
      // This is a data bit
      if (cell) {
        const isTopLeftEye = cellIndex <= 7 && rowIndex <= 7
        const isTopRightEye = cellIndex >= cells.length - 7 && rowIndex <= 7
        const isBottomLeftEye = cellIndex <= 7 && rowIndex >= row.length - 7
        const isEye = isTopLeftEye || isTopRightEye || isBottomLeftEye

        // Add an extra 1 to the top/left so the border isn't cut off by the edge of the canvas
        const cellInfo = {
          color: fgColor,
          left: cellIndex * cellSize + 1,
          top: rowIndex * cellSize + 1,
          size: cellSize,
        }

        // Round the edges of eye bits
        if (isEye) {
          let connectionType = ConnectionType.none

          if (rowIndex > 0 && cells[rowIndex - 1][cellIndex]) {
            connectionType = connectionType | ConnectionType.top
          }

          if (rowIndex < cells.length - 1 && cells[rowIndex + 1][cellIndex]) {
            connectionType = connectionType | ConnectionType.bottom
          }

          if (cellIndex > 0 && cells[rowIndex][cellIndex - 1]) {
            connectionType = connectionType | ConnectionType.right
          }

          if (cellIndex < row.length - 1 && cells[rowIndex][cellIndex + 1]) {
            connectionType = connectionType | ConnectionType.left
          }

          const isInnerEyeX =
            (cellIndex >= 2 && cellIndex <= 4) || (cellIndex >= cells.length - 7 && cellIndex <= cells.length)
          const isInnerEyeY =
            (rowIndex >= 2 && rowIndex <= 4) || (rowIndex >= cells.length - 5 && rowIndex <= cells.length - 3)
          const isInnerEye = isInnerEyeX && isInnerEyeY

          const eyePart = isInnerEye ? EyePart.inner : EyePart.outer

          makeEyeBit(ctx, cellInfo, connectionType, eyePart)
        } else {
          makeDot(ctx, cellInfo)
        }
      }
    })
  })

  // If hideLogo is true then don't render any logo
  // If logoImage is not set default to BloomLogo with colors matching the rest of the QR code
  // Otherwise display the provided logo
  if (!options.hideLogo) {
    const logoImage =
      options.logoImage === undefined ? BloomLogo.getLogo({fgColor: fgColor, bgColor: bgColor}) : options.logoImage

    const image = new Image()
    image.onload = () => {
      const numberOfCellsToCover = Math.floor(cells.length * 0.2)
      const addExtra = numberOfCellsToCover % 2 === 0
      const defaultWidth = numberOfCellsToCover * cellSize + (addExtra ? cellSize : 0)
      const dwidth = options.logoWidth || defaultWidth
      const dheight = options.logoHeight || dwidth
      // Add 1 to accomodate for the 1 shift of the cells
      const dx = (size - dwidth) / 2 + 1
      const dy = (size - dheight) / 2 + 1
      image.width = dwidth
      image.height = dheight

      ctx.save()
      ctx.globalAlpha = options.logoOpacity || 1
      ctx.drawImage(image, dx, dy, dwidth, dheight)
      ctx.restore()
    }
    image.src = logoImage
  }
}

const renderRequestQRCode = (container: HTMLElement, data: RequestData, options: Partial<Options>) => {
  const id = generateId()

  const canvas = document.createElement('canvas')
  canvas.id = id

  drawCanvas(canvas, data, options)

  container.append(canvas)

  return {
    update: updateRequestQRCode(id),
    remove: removeRequestQRCode(id),
  }
}

const updateRequestQRCode = (id: string) => (data: RequestData, options: Partial<Options>) => {
  const canvas = document.querySelector<HTMLCanvasElement>(`#${id}`)

  if (!canvas) return

  drawCanvas(canvas, data, options)
}

const removeRequestQRCode = (id: string) => () => {
  const canvas = document.querySelector(`#${id}`)

  if (canvas) canvas.remove()
}

export {renderRequestQRCode}
