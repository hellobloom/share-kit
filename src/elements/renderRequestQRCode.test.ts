import {renderRequestQRCode} from './renderRequestQRCode'
import {Action, RequestData} from '../..'
import {QROptions} from '../types'

jest.mock('./utils', () => {
  return {
    ...jest.requireActual('./utils'),
    generateId: () => 'bloom-request-element-',
  }
})

describe('renderRequestQRCode', () => {
  let requestQRCode: {update: (data: RequestData, options: Partial<QROptions>) => void; remove: () => void}
  let container: HTMLDivElement

  beforeAll(() => {
    container = document.createElement('div')
  })

  afterAll(() => {
    container.remove()
  })

  beforeEach(() => {
    requestQRCode = renderRequestQRCode(
      container,
      {
        action: Action.attestation,
        token: 'token',
        url: 'https://receive-kit.bloom.co/api/receive',
        org_logo_url: 'https://bloom.co/images/notif/bloom-logo.png',
        org_name: 'Bloom',
        org_usage_policy_url: 'https://bloom.co/legal/terms',
        org_privacy_policy_url: 'https://bloom.co/legal/privacy',
        types: ['phone', 'email'],
      },
      {}
    )
  })

  afterEach(() => {
    requestQRCode.remove()
  })

  test('renders the qr code', () => {
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('removes qr code', () => {
    requestQRCode.remove()

    expect(container.querySelector('canvas')).toBeNull()
  })
})
