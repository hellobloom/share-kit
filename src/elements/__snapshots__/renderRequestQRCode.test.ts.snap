// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderRequestQRCode renders the qr code 1`] = `"<canvas id=\\"bloom-request-element-\\" width=\\"132\\" height=\\"132\\" style=\\"width: 128px; height: 128px; padding: 0%;\\"></canvas>"`;
